### ISSUE TITLE
FileNotFoundError: [Errno 2] No such file or directory

### ISSUE THREAD

**Issue by yohplala** [bmoscon/cryptostore#84]
Hi,

New issue, and this time, added in the right repo :).

Ok, I have started Cryptostore after installing it from Git. It seems to work.
I am using Redis, and configured it to save data to a local binary file 'dump.rdb'(following this tutorial: https://realpython.com/python-redis/)
`dbfilename        dump.rdb`

Ok, so the good thing is that I see this file growing over time.

But, now, when it comes to flushing the data in a `parquet` file, this one does not exist yet, and this seems to be a trouble for cryptostore (started with backfilling plug-in). I don't know for which one it is a trouble: normal cryptostore process or backfilling process.

Any idea?

Here is the log.
```python
2020-06-26 07:32:13,976 : INFO : Starting cryptostore
2020-06-26 07:32:13,977 : INFO : Cryptostore running on PID 13990
2020-06-26 07:32:13,977 : INFO : Backfill running on PID 14000
2020-06-26 07:32:13,979 : INFO : Spawner started
2020-06-26 07:32:13,980 : INFO : Spawner running on PID 14002
2020-06-26 07:32:13,981 : INFO : Aggregator started
2020-06-26 07:32:13,982 : INFO : Aggregator running on PID 14003
2020-06-26 07:32:13,983 : INFO : Cryptostore started
2020-06-26 07:32:13,998 : INFO : start: ['BINANCE'] stop: []
2020-06-26 07:32:13,999 : INFO : message: {"op": "start", "exchange": "BINANCE", "collector": {"channel_timeouts": {"l2_book": 30, "trades": 120}, "retries": 10, "l2_book": {"symbols": ["BTC-USDT", "ETH-USDT", "XRP-USDT"], "book_delta": true, "book_interval": 1440}, "trades": ["BTC-USDT", "ETH-USDT", "XRP-USDT"]}, "config": {"cache": "redis", "redis": {"ip": "127.0.0.1", "port": 6379, "socket": null, "del_after_read": true, "retention_time": null, "start_flush": true}, "storage": ["parquet"], "parquet": {"file_format": ["exchange", "data_type", "pair", "timestamp"], "path": "/home/pierre/Documents/code/cryptostore/data"}, "storage_interval": 300, "plugins": {"backfill": {"module": ["cryptostore.plugin.backfill", "Backfill"], "config": "config.yaml"}}, "backfill": {"BINANCE": {"BTC-USDT": {"start": "2017-01-01"}, "ETH-USDT": {"start": "2017-01-01"}, "XRP-USDT": {"start": "2017-01-01"}}}}}
2020-06-26 07:32:14,000 : INFO : Starting {'op': 'start', 'exchange': 'BINANCE', 'collector': {'channel_timeouts': {'l2_book': 30, 'trades': 120}, 'retries': 10, 'l2_book': {'symbols': ['BTC-USDT', 'ETH-USDT', 'XRP-USDT'], 'book_delta': True, 'book_interval': 1440}, 'trades': ['BTC-USDT', 'ETH-USDT', 'XRP-USDT']}, 'config': {'cache': 'redis', 'redis': {'ip': '127.0.0.1', 'port': 6379, 'socket': None, 'del_after_read': True, 'retention_time': None, 'start_flush': True}, 'storage': ['parquet'], 'parquet': {'file_format': ['exchange', 'data_type', 'pair', 'timestamp'], 'path': '/home/pierre/Documents/code/cryptostore/data'}, 'storage_interval': 300, 'plugins': {'backfill': {'module': ['cryptostore.plugin.backfill', 'Backfill'], 'config': 'config.yaml'}}, 'backfill': {'BINANCE': {'BTC-USDT': {'start': '2017-01-01'}, 'ETH-USDT': {'start': '2017-01-01'}, 'XRP-USDT': {'start': '2017-01-01'}}}}}
2020-06-26 07:32:14,003 : INFO : Collector for BINANCE running on PID 14005
2020-06-26 07:32:14,007 : INFO : Flushing cache
2020-06-26 07:32:14,065 : INFO : Reading BINANCE-l2_book-BTC-USDT
2020-06-26 07:32:14,066 : INFO : No data for BINANCE-l2_book-BTC-USDT
2020-06-26 07:32:14,066 : INFO : Reading BINANCE-l2_book-ETH-USDT
2020-06-26 07:32:14,067 : INFO : No data for BINANCE-l2_book-ETH-USDT
2020-06-26 07:32:14,067 : INFO : Reading BINANCE-l2_book-XRP-USDT
2020-06-26 07:32:14,068 : INFO : No data for BINANCE-l2_book-XRP-USDT
2020-06-26 07:32:14,068 : INFO : Reading BINANCE-trades-BTC-USDT
2020-06-26 07:32:14,069 : INFO : No data for BINANCE-trades-BTC-USDT
2020-06-26 07:32:14,069 : INFO : Reading BINANCE-trades-ETH-USDT
2020-06-26 07:32:14,070 : INFO : No data for BINANCE-trades-ETH-USDT
2020-06-26 07:32:14,070 : INFO : Reading BINANCE-trades-XRP-USDT
2020-06-26 07:32:14,070 : INFO : No data for BINANCE-trades-XRP-USDT
2020-06-26 07:37:14,076 : INFO : Reading BINANCE-l2_book-BTC-USDT
2020-06-26 07:37:14,094 : INFO : l2_book-BINANCE-BTC-USDT: Read 2883 messages from Redis
2020-06-26 07:37:14,361 : ERROR : Aggregator running on PID 14003 died due to exception
Traceback (most recent call last):
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/aggregator/aggregator.py", line 96, in loop
    store.write(exchange, dtype, pair, time.time())
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/data/storage.py", line 37, in write
    s.write(exchange, data_type, pair, timestamp)
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/data/parquet.py", line 86, in write
    pq.write_table(self.data, file_name)
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/parquet.py", line 1343, in write_table
    **kwargs) as writer:
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/parquet.py", line 436, in __init__
    sink = self.file_handle = filesystem.open(path, 'wb')
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/filesystem.py", line 244, in open
    return open(path, mode=mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/me/Documents/code/cryptostore/data/BINANCE-l2_book-BTC-USDT-1593149834.parquet'
Task exception was never retrieved
future: <Task finished coro=<Aggregator.loop() done, defined at /home/me/.local/lib/python3.7/site-packages/cryptostore/aggregator/aggregator.py:42> exception=FileNotFoundError(2, 'No such file or directory')>
Traceback (most recent call last):
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/aggregator/aggregator.py", line 96, in loop
    store.write(exchange, dtype, pair, time.time())
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/data/storage.py", line 37, in write
    s.write(exchange, data_type, pair, timestamp)
  File "/home/me/.local/lib/python3.7/site-packages/cryptostore/data/parquet.py", line 86, in write
    pq.write_table(self.data, file_name)
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/parquet.py", line 1343, in write_table
    **kwargs) as writer:
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/parquet.py", line 436, in __init__
    sink = self.file_handle = filesystem.open(path, 'wb')
  File "/home/me/.local/lib/python3.7/site-packages/pyarrow/filesystem.py", line 244, in open
    return open(path, mode=mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/me/Documents/code/cryptostore/data/BINANCE-l2_book-BTC-USDT-1593149834.parquet'
```

I copy/paste my Cryptostore conffiguration file in case it can help as well.

```python
# Cache
cache: redis
redis:
    ip: '127.0.0.1'
    port: 6379
    socket: null
    del_after_read: true
    retention_time: null
    start_flush: true

# Data sources and data types
exchanges:
    BINANCE:
        channel_timeouts:
            l2_book: 30
            trades: 120
        retries: 10
        l2_book:
            symbols: [BTC-USDT, ETH-USDT, XRP-USDT]
            book_delta: true
            book_interval: 1440
        trades: [BTC-USDT, ETH-USDT, XRP-USDT]

# Storage
storage: [parquet]
parquet:
    file_format: [exchange, data_type, pair, timestamp]
    path: /home/me/Documents/code/cryptostore/data

storage_interval: 300

# Cryptostore Plugin Interface
plugins:
    backfill:
        module: [cryptostore.plugin.backfill, Backfill]
        config: config.yaml

backfill:
    BINANCE:
        BTC-USDT:
            start: '2017-01-01'
        ETH-USDT:
            start: '2017-01-01'
        XRP-USDT:
            start: '2017-01-01'
```

Thanks for any help!

**Comment by yohplala** [comment]
Hi, I made some more testing this evening.
I removed from the config file all lines regarding backfill plug-in, and when starting Cryptostore, it doesn't tell me anything about backefill, so I assume it has not been started.
But the error message still appear.

Further, I created then manually, before starting cryptostore, the folder 'data', as mentionned in the error message:
``` FileNotFoundError: [Errno 2] No such file or directory: '/home/me/Documents/code/cryptostore/data/BINANCE-l2_book-BTC-USDT-1593202927.parquet'
```

and magic! no more error message!
So I guess, somewhere in the code, a check is missing to create if not existing the 'data' folder.
Should be easy to solve if knowing where to look ;).
Bests,

**Comment by bmoscon** [comment]
you specified the path:

/home/me/Documents/code/cryptostore/data

if it doesn't exist, it will complain

**Comment by yohplala** [comment]
ok, crystal clear!
well, maybe it could create the missing directories.
all good now, thanks!

**Comment by bmoscon** [comment]
sure, it could make them if they dont exist, but then if there is a typo in the dir name, the data gets put in a weird place. the documentation could be improved, and the error message could be improved too, so let's re-open this to address those, at a minimum

**Comment by bmoscon** [comment]
I updated the config to make it more clear the path will not be created
